### PR TITLE
Adding Initial Oracle Support

### PR DESCRIPTION
**Pull request by EXONER4TED**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area cmd

/area pkg

> /area docs

> /area tests

**What this PR does / why we need it**:

Adding in Oracle support 🙂
Kernel Crawler supports it, but I realized Driverkit doesn't. I don't think it will be too hard to add.

**Which issue(s) this PR fixes**:

Fixes # N/A

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

Adds `--target oracle`

```release-note
Initial Oracle Support!
```

### CODE REVIEW

**Comment by EXONER4TED** [comment]
Okay, this is just initial support, and copying the standard `rpm` distro template shell script worked pretty well here.

So far, testing is good:
```
╰─➤  _output/bin/driverkit docker --output-module _output/test-oracle.ko --target oracle --kernelrelease 4.18.0-147.0.2.el8_1.x86_64
INFO driver building, it will take a few seconds   processor=docker
INFO kernel module available                       path=_output/test-oracle.ko

╰─➤  _output/bin/driverkit docker --output-module _output/test-oracle.ko --target oracle --kernelrelease 5.4.17-2011.3.2.1.el8uek.x86_64
INFO driver building, it will take a few seconds   processor=docker
INFO kernel module available                       path=_output/test-oracle.ko

╰─➤  _output/bin/driverkit docker --output-module _output/test-oracle.ko --target oracle --kernelrelease 3.10.0-1062.1.1.el7.x86_64
INFO driver building, it will take a few seconds   processor=docker
INFO kernel module available                       path=_output/test-oracle.ko
```
I tried to just grab random 3.x, 4.x, and 5.x kernels to test. Oracle also has these `uek` kernels, and those seem to work fine too.

It's not 100% perfect, I did find one that misses:
```
╰─➤  _output/bin/driverkit docker --output-module _output/test-oracle.ko --target oracle --kernelrelease 4.14.35-2047.510.5.5.el7uek.x86_64
INFO driver building, it will take a few seconds   processor=docker
FATA exiting                                       error="Error: No such container:path: 0327ca8dfb8e0cfbbec44bd116269f44385d51b566e2ca23dd6d3fa75e000d29:/tmp/driver/module.ko"
```
Flipping to debug mode, the error is a bit over my head...
```
DEBU + make CC=/usr/bin/gcc-8.0.0 KERNELDIR=/tmp/kernel
DEBU *make -C /tmp/kernel M=/tmp/driver modules
DEBU *make[1]: Entering directory '/tmp/kernel'
DEBU   CC [M]  /tmp/driver/main.o
DEBU -  CC [M]  /tmp/driver/dynamic_params_table.o
DEBU &  CC [M]  /tmp/driver/fillers_table.o
DEBU $  CC [M]  /tmp/driver/flags_table.o
DEBU #  CC [M]  /tmp/driver/ppm_events.o
DEBU $  CC [M]  /tmp/driver/ppm_fillers.o
DEBU �/tmp/driver/ppm_fillers.c: In function 'f_sys_send_x':
DEBU /tmp/driver/ppm_fillers.c:2412:16: warning: unused variable 'sent_data_pointer' [-Wunused-variable]
DEBU   unsigned long sent_data_pointer;
DEBU                 ^~~~~~~~~~~~~~~~~
DEBU $  CC [M]  /tmp/driver/event_table.o
DEBU &  CC [M]  /tmp/driver/syscall_table.o
DEBU $  CC [M]  /tmp/driver/ppm_cputime.o
DEBU   CC [M]  /tmp/driver/ppm_tp.o
DEBU   LD [M]  /tmp/driver/falco.o
DEBU   Building modules, stage 2.
DEBU   MODPOST 1 modules
DEBU &  SDTINF  /tmp/driver/falco.sdtinfo.c
DEBU �awk: line 110: function strtonum never defined
DEBU awk: line 110: function strtonum never defined
DEBU awk: line 110: function strtonum never defined
DEBU awk: line 110: function strtonum never defined
DEBU awk: line 110: function strtonum never defined
DEBU �make[2]: *** [scripts/Makefile.modpost:135: /tmp/driver/falco.sdtinfo.c] Error 2
DEBU make[2]: *** Deleting file '/tmp/driver/falco.sdtinfo.c'
DEBU )make[1]: Leaving directory '/tmp/kernel'
DEBU .make[1]: *** [Makefile:1622: modules] Error 2
DEBU $make: *** [Makefile:7: all] Error 2
DEBU log pipe close                                error=EOF
```

BUT for a first pass, the vast majority of my testing has come up good. I think what I'd like to do is get this merged through, run it in our loop, and see if I get test against a wider swathe of kernels to see if these errors are widespread or one-offs.

Then come back and fix those issues in a separate MR.

**Comment by FedeDP** [comment]
Hi Logan! Thanks for this :)

> Flipping to debug mode, the error is a bit over my head...

Mmmh my google-fu tells that this is an awk related issue. Can you try to install `gawk` in the builder image being used?

**Comment by EXONER4TED** [comment]
More different errors haha:
```
+ rm -f /tmp/driver/.ctf/driver-extmod.ctf.filelist.raw
+ [[ -n /tmp/driver/.ctf ]]
sh: 1: [[: not found
+ echo /tmp/driver/falco.o
+ rm -f /tmp/driver/.ctf/driver-extmod.ctf.filelist
+ + xargsread -r -r namecolon type file

xargs: file: No such file or directory
+ set -e
+ echo   CTF
  CTF
+ scripts/dwarf2ctf/dwarf2ctf /tmp/driver/.ctf -e /tmp/driver/.ctf/driver-extmod.ctf.filelist
scripts/dwarf2ctf/dwarf2ctf: error while loading shared libraries: libdtrace-ctf.so.1: cannot open shared object file: No such file or directory
make[2]: *** [scripts/Makefile.modpost:252: /tmp/driver/.ctf/driver-extmod.stamp] Error 127
make[1]: *** [Makefile:1622: modules] Error 2
make[1]: Leaving directory '/tmp/kernel'
make: *** [Makefile:7: all] Error 2
```

**Comment by EXONER4TED** [comment]
Aaah... I maybe see what it is... `libdtrace-ctf` is required by Oracle it seems for some UEK builds. It's not immediately obvious that this package exists in debian... let me see if that's possible.

**Comment by EXONER4TED** [comment]
Screenshot 2023-03-23 at 8 52 22 AM
...and it looks like getting `libdtrace-ctf` is not really a thing anymore? So maybe this kernel is just old and no longer matters?

**Comment by FedeDP** [comment]
I agree, it seems like an issue with an old kernel; let's skip it!
> But there is another problem. The builder image no longer builds unless I comment out gcc 4.9 :/ - It looks like Debian quit hosting a jessie repo:

This is much more worrysome :/

File: docker/builders/builder-any-x86_64_gcc8.0.0_gcc6.0.0_gcc5.0.0_gcc4.9.0_gcc4.8.0.Dockerfile [hunk 1]
```diff
@@ -123,7 +125,7 @@ RUN curl -L -o binutils_2.30-22_${TARGETARCH}.deb https://download.falco.org/dep
 
 # Properly create soft link
 RUN ln -s /usr/bin/gcc-4.8 /usr/bin/gcc-4.8.0
-RUN if [ "$TARGETARCH" = "amd64" ] ; then ln -s /usr/bin/gcc-4.9 /usr/bin/gcc-4.9.0; fi;
+# RUN if [ "$TARGETARCH" = "amd64" ] ; then ln -s /usr/bin/gcc-4.9 /usr/bin/gcc-4.9.0; fi;
```
Comment (FedeDP) [inline docker/builders/builder-any-x86_64_gcc8.0.0_gcc6.0.0_gcc5.0.0_gcc4.9.0_gcc4.8.0.Dockerfile:128]: We need 4.9.0 i think; it fixed the build on some kernels (i can't recall which ones though!)
We should find a way to fix gcc4.9.0 installation :(

Comment (EXONER4TED) [reply docker/builders/builder-any-x86_64_gcc8.0.0_gcc6.0.0_gcc5.0.0_gcc4.9.0_gcc4.8.0.Dockerfile:128]: Yes I agree. Also my bad leaving that commented haha. Meant to uncomment.

**Comment by EXONER4TED** [comment]
Trying this... https://unix.stackexchange.com/questions/334888/how-to-install-g-4-9-on-debian-stretch

**Comment by EXONER4TED** [comment]
No luck there... Debian appears to be removing these older packages :/

**Comment by FedeDP** [comment]
Oh god, that means compiling gcc 4.9.3 on the fly :fearful:

**Comment by FedeDP** [comment]
Yay! Great effort!

On a side note, i am going to open an issue to port our current builder to use `gcc` images as base images: https://hub.docker.com/_/gcc
They are multiplatform (also available for arm32 and s390x, if driverkit would ever want to support them), so that we might drop the `arch` field from current image regex, and they are available for each gcc version without any pain for us.

**Comment by FedeDP** [comment]
Changes LGTM! Can you update documentation? (`--target` in docs/driverkit.md, docs/driverkit_docker, ... etc etc)

File: pkg/driverbuilder/builder/oracle.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	"strings"
+
+	"github.com/falcosecurity/driverkit/pkg/kernelrelease"
+)
+
+//go:embed templates/oracle.sh
+var oracleTemplate string
+
+// TargetTypeoracle identifies the oracle target.
+const TargetTypeoracle Type = "oracle"
```
Comment (FedeDP) [inline pkg/driverbuilder/builder/oracle.go:15]: Can we name it `ol` (that is ID from `/etc/os-release` in oraclelinux systems)?
```
✓ (457ms) 12:09:28 [federico][fededp-sysdig][update/oracle][~/Work/kernel-crawler] $ docker run -ti --rm oraclelinux:6 cat /etc/os-release | grep ID | head -n1
ID="ol"
✓ (576ms) 12:09:33 [federico][fededp-sysdig][update/oracle][~/Work/kernel-crawler] $ docker run -ti --rm oraclelinux:7 cat /etc/os-release | grep ID | head -n1
ID="ol"
✓ (617ms) 12:09:39 [federico][fededp-sysdig][update/oracle][~/Work/kernel-crawler] $ docker run -ti --rm oraclelinux:8 cat /etc/os-release | grep ID | head -n1
ID="ol"
✓ (586ms) 12:09:43 [federico][fededp-sysdig][update/oracle][~/Work/kernel-crawler] $ docker run -ti --rm oraclelinux:9 cat /etc/os-release | grep ID | head -n1
ID="ol"
```

Comment (FedeDP) [reply pkg/driverbuilder/builder/oracle.go:15]: See also https://github.com/falcosecurity/kernel-crawler/pull/107

Comment (EXONER4TED) [reply pkg/driverbuilder/builder/oracle.go]: Done!

File: pkg/driverbuilder/builder/oracle.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	if strings.Contains(version, "_") {
+		version = strings.Split(version, "_")[0]
+	}
+
+	// list of possible UEK versions, which are used in the URL - ex: "UEKR3"
+	// may need to evolve over time if Oracle adds more
+	ueks := []string{"R3", "R4", "R5", "R6"}
+
+	// template the kernel info into all possible URL strings
+	urls := []string{
```
Comment (FedeDP) [inline pkg/driverbuilder/builder/oracle.go:58]: Can we add oracle9 support? 
https://github.com/falcosecurity/kernel-crawler/blob/main/kernel_crawler/oracle.py#L32

(and also `ueks R7`!)

Comment (EXONER4TED) [reply pkg/driverbuilder/builder/oracle.go:58]: 🤦 hahaha I totally just forgot about it. 2 sec

Comment (EXONER4TED) [reply pkg/driverbuilder/builder/oracle.go:58]: Added!

**Pull request merged**